### ISSUE TITLE
Is there anyway to create html log/report when script is stopped by human?

### ISSUE THREAD

**Issue by Pantyhose-S** [nokia/RED#170]
I have a roaster test script.
The testcase will never be finished by it self (infinite loop), I will run it for the whole weekend and stop it at monday by click the "stop" button.
When this happens, there is no html log, only message view.
Is there any way to create html log and report?

**Comment by adam-wypych** [comment]
If you have output.xml and log.xml you can try to run Robot Framework internal tools in following order:
1) (FixXml) https://bitbucket.org/robotframework/fixml : _Tool for fixing Robot Framework output files that are broken._
2) (Reboot) http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#post-processing-outputs : _Tool for generating logs and reports based on XML outputs and for combining multiple outputs together._

**Comment by michalanglart** [comment]
Shortly: we need to take a look but it should be achievable on unix-like systems, may be harder with windows.

Basically Robot handles gracefuly termination: http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#stopping-test-execution-gracefully however now with RED it does not work.

There are basically two things happening when user clicks 'Terminate' button:

1) We're sending a message to our agent which is hooked into the execution (http://nokia.github.io/RED/help/user_guide/launching/red_agent.html) and it will call python's sys.exit() after current keyword or before next one thus killing the execution process.
2) At the same time RED calls java's Process#destroy() on process handle which on linux sends SIGTERM signal to that process.

Putting this altogether: if we would not send TERMINATE message to agent all the output/logs/report should be properly generated assuming that this is some linux system.

As a side note: we still need to send this TERMINATE message in remote execution since we do not have a handle to running process.

I don't know however if we can also achieve this with Windows.

**Comment by michalanglart** [comment]
Hi,

no we don't know any workaround as of now. We'll need to dig into process termination code and change something to make it work.

**Comment by adam-wypych** [comment]
I think we can borrow logic from RIDE:

https://github.com/robotframework/RIDE/blob/master/src/robotide/lib/robot/running/signalhandler.py
https://github.com/robotframework/RIDE/blob/master/src/robotide/contrib/testrunner/testrunner.py

**Comment by apallier** [comment]
@michalanglart I don't know if what I will say is correct (I don't know your project enough...) but I see in [WindowsProcessTreeHandler.java](https://github.com/nokia/RED/blob/f4499ffdc753e67ec75abbfaaafad99c4e236b3e/src/RobotFrameworkCore/org.robotframework.ide.core-functions/src/main/java/org/rf/ide/core/jvmutils/process/WindowsProcessTreeHandler.java#L55) and in [UnixProcessTreeHandler.java](https://github.com/nokia/RED/blob/f4499ffdc753e67ec75abbfaaafad99c4e236b3e/src/RobotFrameworkCore/org.robotframework.ide.core-functions/src/main/java/org/rf/ide/core/jvmutils/process/UnixProcessTreeHandler.java#L46) that the command line to kill a process is a little bit "brutal":
```
// Windows
wmic process where ProcessID=... delete
```
```
// Linux: 
kill -9
```
Maybe you should send a CTRL-C signal instead?

**Comment by adam-wypych** [comment]
@apallier This functionality was implemented as workaround for lack of kill process tree in Java 6, 7 and 8. As far as I remember, you can stop Jython process but still child process of Java will be running as zombie.

**Comment by michalanglart** [comment]
Yes, moreover this workaround is used only to kill the xml-rpc server which RED is running in background. Tests termination is done through eclipse code: https://github.com/eclipse/eclipse.platform.debug/blob/master/org.eclipse.debug.core/core/org/eclipse/debug/core/model/RuntimeProcess.java 

The `terminate()` method calls Java `Process#destroy()` method in which (as javadoc states): 

>  Whether the subprocess represented by this Process object is forcibly terminated or not is implementation dependent

**Comment by adam-wypych** [comment]
@michalanglart I didn't check but maybe with calling https://github.com/robotframework/RIDE/blob/master/src/robotide/lib/robot/running/signalhandler.py from `TestRunnerAgent` we can stop running test before call `terminate()`.

**Comment by michalanglart** [comment]
At first sight it looks that running this `signalhandler` in separate thread + overriding `terminate()` method properly could be promising. We need to experiment with this a little bit.

**Comment by apallier** [comment]
@michalanglart If you have a pre-release with a correction, I will be interest to test it :)

**Comment by apallier** [comment]
Hi, could you say me if you intend to fix this issue? And if yes, when? :)

**Comment by michalanglart** [comment]
Hi,

yes, we do intend to provide a fix. Probably not the next release but hopefully version 0.8.4+2 is feasible.

**Comment by adam-wypych** [comment]
@michalanglart I think @apallier wants to ask about more concrete time. I don't see any release plan with approximately time on RED github page.

// cc @KrzysztofJozefowicz

**Comment by apallier** [comment]
@michalanglart Happy to know that you intend to shortly fix this issue! Thank you.

**Comment by KrzysztofJozefowicz** [comment]
We are the middle of development of 0.8.5 which should be released in ~2 weeks (2nd part of April), we will take #170 into analysis and development for 0.8.6 and should be ready around a month or so (2nd part of May). If we would have pre-release fix, we will share it with you for testing.

**Comment by michalanglart** [comment]
Soft termination was released in 0.8.5

It is done through separate button in Console view. On unix based machines it sends SIGINT by calling `kill -2 PID`. On windows machines it runs custom python script which uses `kernel32.dll` in order to attach to the console of running process and send `Ctrl+C` to it.
